numerics/cbrt.cpp

```diff
@@ -3,35 +3,166 @@
+    return {g₁, g₂, g₃, g₄};
+  } else {
+    auto const [w, g₄] = TwoSum(b₂, d₂);
+    auto const [u, v] = TwoSum(d₁, b₁);
+    auto const [x, g₃] = TwoSum(v, w);
+    auto const [g₁, g₂] = TwoSum(u, x);
+    return {g₁, g₂, g₃, g₄};
+  }
+}
+
+// See [Nie04], algorithm 8; note that we use 0-based indices where Nievergelt
+// uses 1-based indices.  In order to avoid allocations, we do not return a
+// vector, but instead an array of the maximum possible size.  Should it be
+// necessary to return the number of nonzero elements of the result, some sort
+// of bounded vector would probably be useful.  Note that this size is k in
+// Nievergelt’s procedure, but with our indexing it is is k + 1 at the return
+// statement.
+template<std::size_t n>
+std::array<double, n> PriestNievergeltNormalize(std::array<double, n> const f) {
+  std::array<double, n> s{};
+  int k = 0;
+  s[0] = f[0];
+  for (int j = 1; j < n; ++j) {
+    auto const [c, d] = TwoSum(s[k], f[j]);
+    s[k] = c;
+    if (d != 0) {
+      int l = k - 1;
+      k = k + 1;
+      while (l >= 0) {
+        auto const [cʹ, dʹ] = TwoSum(s[l], s[l + 1]);
+        s[l] = cʹ;
+        if (dʹ == 0) {
+          k = k - 1;
+        } else {
+          s[l + 1] = dʹ;
+        }
+        l = l - 1;
+      }
+      s[k] = d;
+    }
+  }
+  return s;
+}
+
+bool CorrectionPossiblyNeeded(double const r₀,
+                              double const r₁,
+                              double const r̃,
+                              double const τ) {
+  // The order of the conjunction matters for branch predictability here.  The
+  // left-hand side is almost always false: it is true only when r = r₀ + r₁ is
+  // very close to a tie (for r̃ ≠ r₀) or to a representable number (for r̃ = r₀).
+  // On the other hand, it is hard to predict whether r̃ = r₀, i.e., whether r is
+  // closer to a representable number than to a tie.
+  return std::abs(0.5 * (r̃ - r₀) - r₁) <= τ * r₀ && r̃ != r₀;
+}
+
+double CorrectLastBit(double const y, double const r₀, double const r̃) {
+  double const a = std::min(r₀, r̃);
+  double const b = 0.5 * (std::max(r₀, r̃) - a);
+  return CbrtOneBit(y, a, b) ? std::max(r₀, r̃) : a;
+}
+
+bool CbrtOneBit(double y, double a, double b) {
+  double const b² = b * b;
+  double const b³ = b² * b;
+  DoublePrecision<double> const a² = TwoProduct(a, a);
+  auto const& [a²₀, a²₁] = a²;
+  DoublePrecision<double> const a³₀ = TwoProduct(a²₀, a);
+  DoublePrecision<double> minus_a³₁ = TwoProduct(a²₁, -a);
+  auto const& [a³₀₀, a³₀₁] = a³₀;
+  // In cbrt.pdf, where we are specifically considering the computation of the
+  // 54th bit, ρ is referred to as ρ₅₃, and ρ_next as ρ₅₄ˌ₁.
+  // ρ = y - a³ = y - a³₀ - a³₁ = y - a³₀₀ - a³₀₁ - a³₁;
+  double const ρ₀ = y - a³₀₀;  // Exact.
+  // ρ = ρ₀ - a³₀₁ - a³₁;
+  std::array<double, 4> const ρ = PriestNievergeltNormalize(
+      NievergeltQuadruplyCompensatedStep(TwoDifference(ρ₀, a³₀₁), minus_a³₁));
+  CHECK_EQ(ρ[3], 0);
```

Comment:
That shouldn’t matter for performance since this is the slow path ; since we test more values in release than debug, it might matter for test coverage.